### PR TITLE
Use SMC to initialize compartmental models

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 

This replaces custom `.heuristic()` methods with a default generic `SMCFilter`-based initialization strategy, greatly simplifying the process of creating new compartmental models. Users can still override this default `.heuristic()` if needed.

:tada: hooray for composable inference :tada:

## Tested
- [x] refactoring covered by existing tests
- [x] ran `python sir.py -p 10000 -d 60 -f 30 --verbose --plot` and confirmed good posteriors and mixing:
```
         mean       std    median      5.0%     95.0%     n_eff     r_hat
 R0      1.48      0.08      1.47      1.36      1.61     16.89      1.09
rho      0.48      0.03      0.49      0.44      0.53     29.28      1.14
```
image

### CODE REVIEW

File: pyro/contrib/epidemiology/compartmental.py [hunk 1]
```diff
@@ -433,3 +456,45 @@ def enum_shape(position):
...
+            if site["type"] == "sample":
+                state[name] = site["value"]
+
+        self.t = 0
+        state.update(self.model.initialize(params))
+        self.step(state)  # Take one step since model.initialize is deterministic.
+
+    def step(self, state):
+        with poutine.block(), poutine.condition(data=state):
+            params = self.model.global_model()
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/compartmental.py:482]: why is this needed exactly?

Comment (fritzo) [reply pyro/contrib/epidemiology/compartmental.py:482]: Recall that `SMCFilter` tracks all state in a `state` dict and periodically resamples all that state. In this use case the `state` dict contains values of all sample sites that we will eventually replay. In particular it contains sample sites for global variables. Now since resampling changes those sites, and since `global_model()` not only draws samples but performs computation on those samples, we need to re-perform that computation after each resampling. Hence we block (to avoid this being traced again), and condition on the `SMCFilter`-managed `state` to get the latest resampled `params` each step.
